packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts

```diff
@@ -33,32 +42,55 @@ export type InsightsProps<
   onEvent?: (event: InsightsEvent, insightsClient: TInsightsClient) => void;
 };
 
+const ALGOLIA_INSIGHTS_SRC =
+  'https://cdn.jsdelivr.net/npm/search-insights@2.3.0/dist/search-insights.min.js';
+
 export type CreateInsightsMiddleware = typeof createInsightsMiddleware;
 
 export function createInsightsMiddleware<
   TInsightsClient extends null | InsightsClient
->(props: InsightsProps<TInsightsClient>): InternalMiddleware {
+>(props: InsightsProps<TInsightsClient> = {}): InternalMiddleware {
   const {
     insightsClient: _insightsClient,
     insightsInitParams,
     onEvent,
-  } = props || {};
+  } = props;
+
+  let insightsClient: InsightsClient = _insightsClient || noop;
+
   if (_insightsClient !== null && !_insightsClient) {
-    if (__DEV__) {
-      throw new Error(
-        "The `insightsClient` option is required if you want userToken to be automatically set in search calls. If you don't want this behaviour, set it to `null`."
-      );
-    } else {
-      throw new Error(
-        'The `insightsClient` option is required. To disable, set it to `null`.'
-      );
-    }
+    safelyRunOnBrowser(({ window }: { window: any }) => {
+      const pointer = window.AlgoliaAnalyticsObject || 'aa';
+
+      if (typeof pointer === 'string') {
+        insightsClient = window[pointer];
+      }
+
+      if (!insightsClient) {
+        window.AlgoliaAnalyticsObject = pointer;
+        window[pointer] =
+          window[pointer] ||
+          function (...args: any[]) {
+            (window[pointer].queue = window[pointer].queue || []).push(args);
+          };
+
+        insightsClient = window[pointer];
+
+        try {
+          const script = document.createElement('script');
+          script.async = true;
+          script.src = ALGOLIA_INSIGHTS_SRC;
+          script.onerror = () => {
+            // @TODO: it would be useful to track errors
+          };
+          document.body.appendChild(script);
+        } catch (e) {
+          // @TODO: it would be useful to track errors
+        }
```

Comment:
alternative that could be interesting, so it can be at least seen in the console

```suggestion
          script.onerror = () => {
            throw new Error('.... insights failed to load ...');
          };
          document.body.appendChild(script);
        } catch (e) {
          throw new Error('.... insights failed to load ...');
        }
```

Reply:
with some reorganisation, the middleware could likely also be found to be removed, but not sure if that's right to do?

Reply:
> alternative that could be interesting, so it can be at least seen in the console

What would we recommend the customer to do in this case? Redirect them to the installation guide part around Insights? In dev environment it's unlikely they would have CSP headers set so I don't know what is the guidance we can provide them by showing an error.

Reply:
Yes, there's quite a few different reasons why it could fail, so the error message wouldn't be easy to write

Reply:
I would go with warnings, not errors, as we shouldn't break the app for not being able to load Insights, no?

A warning in dev mode with a message like this:

>InstantSearch wasn't able to automatically load the Insights client. This can happen for a number of reasons; please check our troubleshooting guide for more info: https://algolia.com/doc/...

Reply:
an error in an async context doesn't break the page, but warning could also work.

Do we know what to suggest to people though?

Reply:
dev-only has a downside that the most likely situation it would fail in is production (due to stricter CSP setup than localhost usually). That's why I was more thinking about errors

Reply:
The use cases we know about are:
- CSP, they need to accept jsDelivr or manually load Insights + set up the middleware
- jsDelivr is down, they can manually load Insights + set up the middleware for 100% success

And we could still send warnings/errors in production indeed. Maybe errors are better since it could be reported by Sentry or Bugsnag if they use it?